### ISSUE TITLE
Change `no-duplicate-case` rule to comparing tokens

### ISSUE THREAD

**Issue by ota-meshi** [eslint/eslint#13485]
**What rule do you want to change?**

- `no-duplicate-case`

**Does this change cause the rule to produce more or fewer warnings?**

Produce more warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**

Makes the process of detecting duplicates compare by token, like `no-dupe-else-if`.

**Please provide some example code that this change will affect:**

```js
switch (a) {
  case a.b:
  case a.b: // Already reported.
  case a.      b: // Will be reported.
  case a./**/b: // Will be reported.
}
```

**What does the rule currently do for this code?**

The comparison is text based.

**What will the rule do after it's changed?**

`a.b`, `a.    b` and `a./**/b` are reported as duplicates.

**Are you willing to submit a pull request to implement this change?**

If there is consensus, I will submit a pull request.

---

refs https://github.com/eslint/eslint/issues/13466#issuecomment-654867869

**Comment by mdjermanovic** [comment]
@ota-meshi thanks for the issue!

I'll champion this.

The only downside is that this increases time complexity (currently used `Set` methods are sublinear, plus we'll additionally have to compare arrays of tokens), but I don't think it will be an issue since it is expected to work on small lists.

Only, I'm not sure should we make this change in a minor version, or rather wait for the next major version.

**Comment by mdjermanovic** [comment]
There are 3 upvotes from team members, so marking as accepted.

> If there is consensus, I will submit a pull request.

PR is welcome now!

(I think we should discuss if this should be a semver-minor or semver-major change, but that shouldn't affect the PR)

**Comment by ota-meshi** [comment]
@mdjermanovic Thank you for letting me know!

I will work on this PR after today's work.
